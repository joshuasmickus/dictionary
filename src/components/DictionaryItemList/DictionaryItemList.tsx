import * as React from 'react';
import { Link, Redirect } from 'react-router-dom';

import { IDictionaryItem, IDictionaryName, IDictionaryState } from '../../models/common';
import { deleteDictionaryItem } from '../../state/dictionary/dictionaryRoutines';

interface IDictionaryItemListProps {
  items: IDictionaryItem[];
  list: IDictionaryName[];
  deleteDictionaryItemRequest: (dictionaryId: string, id: string) => void;
  match: any;
};

export const DictionaryItemList: React.FC<IDictionaryItemListProps> = ({
  items,
  list,
  deleteDictionaryItemRequest,
  match
}) => {
  const { id: dictionaryId } = match.params;
  const deleteItem = (id: IDictionaryItem['id']) => (event: React.MouseEvent) => {
    event.preventDefault();

    deleteDictionaryItemRequest(dictionaryId, id);
  };
  const listItems = items.filter((item) => item.dictionaryId === dictionaryId);
  const filteredDictionaries = list.filter((dictionary) => dictionary.id === dictionaryId);
  const shouldShowItems = listItems.length > 0;

  if (filteredDictionaries.length === 0) {
    return <Redirect to="/" />;
  }

  const { name: dictionaryName } = filteredDictionaries[0];

  return (
    <React.Fragment>
      <h2>Entries for {dictionaryName}</h2>
      {
        shouldShowItems &&
        <Link to={`/edit-dictionary/${dictionaryId}/add-item`}>Add item</Link>
      }
      <ul>
        {
          shouldShowItems && listItems.map((item: IDictionaryItem) =>
            (
              <li key={item.id}>
                {item.from} {item.to}&nbsp;
                <Link to={`/edit-dictionary/${dictionaryId}/edit-item/${item.id}`}>edit</Link> -&nbsp;
                <a onClick={deleteItem(item.id)} href={`/edit-dictionary/${dictionaryId}/delete-item/${item.id}`}>delete</a>
              </li>
            )
          )
        }
        { !shouldShowItems && <li>There are no dictionary items, <Link to={`/edit-dictionary/${dictionaryId}/add-item`}>add one</Link></li> }
      </ul>
    </React.Fragment>
  );
};

export const mapStateToProps = ({ dictionary }: IDictionaryState) => ({
  items: dictionary.items,
  list: dictionary.list
});

export const mapDispatchToProps = (dispatch: (action: any) => void) => ({
  deleteDictionaryItemRequest: (dictionaryId: IDictionaryName['id'], id: IDictionaryItem['id']) => dispatch(deleteDictionaryItem.request({ dictionaryId, id }))
});
