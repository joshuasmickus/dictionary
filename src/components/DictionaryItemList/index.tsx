import { connect } from 'react-redux';

import {
  DictionaryItemList,
  mapDispatchToProps,
  mapStateToProps
} from './DictionaryItemList';

export default connect(mapStateToProps, mapDispatchToProps)(DictionaryItemList);
