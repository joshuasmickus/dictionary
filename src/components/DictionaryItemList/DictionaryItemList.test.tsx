import { createMemoryHistory } from 'history'
import * as React from 'react';
import { Router } from 'react-router-dom';
import { render } from 'react-testing-library';

import { DictionaryItemList } from './DictionaryItemList';

const id = 'someid';
const requiredProps = {
  deleteDictionaryItemRequest: jest.fn(),
  items: [],
  list: [{ name: 'something', id }],
  match: { params: { id }}
};

describe('Given a DictionaryItemList component', () => {
  describe('when it is rendered', () => {
    it('should match the snapshot', () => {
      const { container } = render(<Router history={createMemoryHistory()}><DictionaryItemList {...requiredProps} /></Router>);

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
