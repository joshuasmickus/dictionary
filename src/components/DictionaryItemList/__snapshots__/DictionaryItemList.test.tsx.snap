// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a DictionaryItemList component when it is rendered should match the snapshot 1`] = `
<h2>
  Entries for 
  something
</h2>
`;
